### ISSUE TITLE
ptr::read() is slower than directly dereferencing

### ISSUE THREAD

**Issue by jrmuizel** [rust-lang/rust#73258]
```rust
#[derive(Clone, Copy)]
pub enum Foo {
    A, B, C, D,
}

pub fn foo(v: *const Foo) -> Foo {
    let k: Option<Foo> = unsafe { Some(*v) };
    return k.unwrap();
}
```

compiles to
```asm
example::foo:
        mov     al, byte ptr [rdi]
        ret
```

but replacing `*v` with `v.read()`

```rust
#[derive(Clone, Copy)]
pub enum Foo {
    A, B, C, D,
}

pub fn foo(v: *const Foo) -> Foo {
    let k: Option<Foo> = unsafe { Some(v.read()) };
    return k.unwrap();
}
```

compiles to:

```asm
example::foo:
        push    rax
        mov     al, byte ptr [rdi]
        cmp     al, 4
        je      .LBB0_1
        pop     rcx
        ret
.LBB0_1:
        lea     rdi, [rip + .L__unnamed_1]
        lea     rdx, [rip + .L__unnamed_2]
        mov     esi, 43
        call    qword ptr [rip + core::panicking::panic@GOTPCREL]
        ud2
```

This came up from #71257

**Comment by alex** [comment]
godbolt for the bad version here: https://rust.godbolt.org/z/CWboVo

took me a second, but the panic is from the unwrap, suggesting that the discriminant range isn't propagated correctly

**Comment by jrmuizel** [comment]
This is presumably because the `copy_nonoverlapping` in the implementation of `ptr::read()` does not emit range metadata.

**Comment by alex** [comment]
I think that's right, looking at the LLVM IR there's no range metadata on the `load`.

I'm a bit confused about what part of the machinery is responsible for turning the `copy_nonoverapping` intrinsic into an LLVM load though, a quick grep of the code seems to suggest that it's emitted as a `memcpy` into LLVM.

Which then got me thinking, shouldn't the load after the memcpy really be where the range annotation is? I don't think I know enough about how LLVM's range metadata works to go further.

**Comment by hanna-kruppe** [comment]
This matches my impression from skimming over `-print-after-all` output, though I'll add that the call to `ptr::read()` *does* have the range metadata, so if that metadata got propagated before or during inlining, we'd probably be in a better position. But a simpler solution in the short term might be to implement `ptr::read` with a compiler intrinsic instead of the current MaybeUninit+copy_nonoverlapping contraption.

**Comment by bugadani** [comment]
It looks like the problem is MaybeUninit: https://rust.godbolt.org/z/bMWrqo

**Comment by bugadani** [comment]
I'm not sure if placing the blame on `ptr::read` is correct here:
https://rust.godbolt.org/z/Yr7K6c

Notice how `bar` just feeds `foo` with a known value and is optimized away. Looks like enum metadata is not attached to input args?

**Comment by scottmcm** [comment]
LLVM-IR diff: https://rust.godbolt.org/z/vMex9ooxq

Looks like the root cause is that `read` doesn't get `!range` metadata:
```diff
-  %_3 = load i8, ptr %v, align 1, !dbg !11, !range !13, !noundef !10
+  %tmp.0.copyload = load i8, ptr %v, align 1, !dbg !11
```

Will hopefully be fixed by #109035

**Comment by scottmcm** [comment]
Test: https://github.com/rust-lang/rust/blob/master/tests/codegen/issues/issue-73258.rs
